ui: Correct responses for restricted access to non-admin users

koorosh opened the pull request:
Resolves #48152, #50360

Depends on https://github.com/cockroachdb/admin-ui-components/pull/37
Depends on https://github.com/cockroachdb/yarn-vendored/pull/44

Previously, request handlers for node locations and eventslogs didn't
check if the user has an admin role or not and proceeds with an attempt to query
data from tables with granted permissions for admin users only. This
error handled as server error then.

Current change validates if a user has admin role before requesting
data from tables and responds with Permission denied error when user
doesn't have an admin role.

UI changes:
- Add InlineAlert component to display redesigned messages;

**Screens**
Single info message (Restricted permissions)
<img width="1174" alt="Screenshot 2020-07-02 at 18 34 03" src="https://user-images.githubusercontent.com/3106437/86460281-c3d7c880-bd30-11ea-9ef4-006e436c26e8.png">

Single error message
<img width="1154" alt="Screenshot 2020-07-02 at 18 41 54" src="https://user-images.githubusercontent.com/3106437/86460294-cd613080-bd30-11ea-9e81-9009f4d47b5e.png">

Multiple error messages
<img width="1259" alt="Screenshot 2020-07-02 at 18 24 34" src="https://user-images.githubusercontent.com/3106437/86460323-d9e58900-bd30-11ea-81bf-6ff47fdadba1.png">

koorosh committed:
ui: messages for restricted access for non-admin users

Previously, request handlers for node locations and eventslogs didn't
check if user has admin role or not and proceeds with attempt to query
data from tables with granted permissions for admin users only. This
error handled as server error then.

Current change validates if user has admin role before requesting
data from tables and responds with Permission denied error when user
doesn't have admin role.

Release note (admin ui change): Show Permission denied error on Node Map
and Events views for non-admin users.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50869)
<!-- Reviewable:end -->

koorosh committed:
ui: Integrate InlineAlerts into Loading component

Loading component is responsible to provide feedback for user whether
data is ready, is loading or failed to load with some error.

Previously, when loading failed, error message was rendered inside of
Loading component. Now, rendering of errors are delegated to
 InlineAlert component.
Loading component is responsible to properly map error data to
InlineAlert props. And it has to take into account next cases:
- single error
- multiple errors with the same intent
- multiple errors with different intent
- Restricted permissions error is handled as special case and has to
be rendered as "info" alert.

Release note (admin ui change): Redesign inline error alerts on pages.
Alerts that shown when user has insufficient rights to see some
resources (for instance non admin user on Databases page).

dhartunian commented on pkg/server/admin.go:
```diff
@@ -1494,12 +1493,13 @@ func (s *adminServer) Locations(
 	}
-
+	if !isAdmin {
+		return nil, errInsufficientPrivilege
```
Wondering if the better thing to do in these cases is find a good way to pass the SQL privilege error through to the UI. Otherwise we get stuck "double checking" all these instances which is both error prone and confusing.

Did you explore that option @koorosh? I think sticking to SQL as "the truth" as much as possible will be good here because the permissions model will change in the future but as long as we use the SQL-level error the UI will always be doing the right thing.

koorosh replied:
Good question. My concern is that we have `role_members` table with an explicit `isAdmin` field to indicate users' privilege. I thought it is the source of truth right now. Will it be changed in the future?

I agree that right we don't handle permission error in a generic way with other possible SQL errors but on the other hand, restricted access for querying tables isn't the only reason for permission errors (for instance, it could be restricted depending on context).

koorosh replied:
@dhartunian since last discussion, there was a change on server which introduced unified validation for admin permissions (2be60c1b) so now it returns "this operation requires admin privilege" error message from server.

current: 
```
This information is not available due to the current value of the 'server.remote_debugging.mode' setting.
```

message from server:
```
RequestError: this operation requires admin privilege
```
@Annebirzin , would it be okay to show server message instead of current one?

Annebirzin replied:
@koorosh yep I think this is actually more clear than the current message

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Annebirzin, @elkmaster, and @koorosh)

___
*[pkg/server/admin.go, line 1501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50869#-MB9j3qTCouZfgwwf9HF-r3-1501:-MKk0OVM52cz48DudQm8:bb4uzs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d6fb82c140d1c6bb9104970d17f39676d594c31/pkg/server/admin.go#L1501)):*
<details><summary><i>Previously, Annebirzin (Anne Birzin) wrote…</i></summary><blockquote>

@koorosh yep I think this is actually more clear than the current message 
</blockquote></details>

Sounds good @koorosh, use the new `requireAdminUser` function in these places.


<!-- Sent from Reviewable.io -->

koorosh committed:
ui: update user permissions validations for adminui endpoints

Current change reuses `requireAdminUser` function to validate
whether user has admin role to request resources or not for
Events endpoint.
In addition, removed unnecessary validation for admin role on
Locations endpoint. This endpoint can be accessed by users without
admin role as well.

Release note: none

koorosh committed:
ui: provide server generated errors for restricted permissions

Initially, Admin UI messages for restricted permissions were custom
and didn't provide enough details.

Now, client consumes and decodes server Response errors and displays
error messages provided by server.

Release note (ui change): Show error messages provided by server
instead of custom generic messages defined on client side. For instance,
messages about permission restrictions to show pages for non-admin roles

koorosh committed:
ui: reuse Loading component from admin-ui-components

Currently, Loading component has been defined in two projects: in pkg/ui
and admin-ui-components which caused code duplication.
admin-ui-components package has updated component styles so it is
reused in Admin UI.
To avoid extra changes across entire project, Loading component is
re-exported from its initial location to preserve the same import
paths where it is used.

Release note (ui change): display error messages provided by
server in alert messages (for instance when permissions denied to
display some page).

dhartunian reviewed (changes requested):
> Reviewed 10 of 10 files at r9, 2 of 2 files at r10, 3 of 3 files at r11, 2 of 2 files at r12, 5 of 5 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Annebirzin, @elkmaster, and @koorosh)

___
*[pkg/ui/src/views/cluster/containers/events/index.tsx, line 181 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/50869#-MMrCN1bDjh2kQK2c9wP:-MMrCN1bDjh2kQK2c9wQ:b-1twitq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a4ac76ef788fdeb49837c39633f822a8da464a1/pkg/ui/src/views/cluster/containers/events/index.tsx#L181)):*
> ```XML
>           loading={!events}
>           error={lastError}
>           render={this.renderContent}
> ```

I think `renderContent` needs to be bound to `this` in this component, otherwise the Events page fails to render. Or change `renderContent` so it doesn't depend on `this.props`. I wasn't able to get it to render without doing that.
___
*[pkg/ui/src/views/shared/components/loading/loading.spec.tsx, line 19 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/50869#-MMrMMqV3MwEyjWlXplV:-MMrMMqV3MwEyjWlXplW:b-9op11i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d01abd67236c0235b47fff29b8dd6bd243d17405/pkg/ui/src/views/shared/components/loading/loading.spec.tsx#L19)):*
> ```XML
> import Loading from "src/views/shared/components/loading";
> import { ReactWrapper, mount } from "enzyme";
> import { InlineAlert } from "oss/src/components/inlineAlert/inlineAlert";
> ```

remove `oss`


<!-- Sent from Reviewable.io -->

koorosh commented:
Updated PR with fix for Events page and also updated imports to use direct imports from `@cockroachlabs/admin-ui-components`

dhartunian reviewed (commented):
Does this need a version bump for `admin-ui-components` to pick up latest `Loading` version?

> Reviewed 22 of 22 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50869#-:-MMvjtY30FtKyxS-gNCE:b-lbo1n2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Annebirzin, @elkmaster, and @koorosh)



<!-- Sent from Reviewable.io -->

koorosh commented:
> Does this need a version bump for `admin-ui-components` to pick up latest `Loading` version?
> 
> >

Correct, bumped `admin-ui-components` to the `0.1.24` version (https://github.com/cockroachdb/admin-ui-components/pull/37)

dhartunian commented:
@koorosh now there's a conflict w/ `yarn-vendor` can you give that an update as well since we've merged other updates recently.

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work on this!

> Reviewed 29 of 29 files at r15, 1 of 2 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50869#-:-MNQPNkT2Ss1SbpMYmaX:b7szsbh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Annebirzin, @dhartunian, @elkmaster, and @koorosh)



<!-- Sent from Reviewable.io -->

koorosh commented:
bors r+
